### PR TITLE
Knives Out

### PR DESCRIPTION
**Pull request by tenmillimaster**
#### Summary
Content "Adds 2 bayonets, 1 fighting knife and adjusts several other knives"

#### Purpose of change

The volumes and weights on several knives were off.
Bayonets were spawning in inappropriate places.
The Damage on some knives were disproportionately high (RM42, combat knife, Kirpan).
The RM42 could have been a real combat knife, not a fictional one.
Additional bayonets could help make the different monsters from each service branch feel different.
No bayonet makes any gun long enough to get reach attack (1.5m MIN, 1.8 preferable).

#### Describe the solution

Perform volume and weight audit of several bayonets and knives. This had the unintentional effect of making most knives faster and increasing DPS. Making most knives faster yet dealing less stab damage seemed appropriate.

Turned the combat knife into the "National Guard bayonet", or the m7(standing in also for the m4-m6 bayonets, and the m3 fighting knife if you pretend it's a bayonet) bayonet, which should be considered one of the best fighting knives available. The national guard would have piles of these to hand out to reservists.

Added the M9 bayonet, as the "Army Bayonet". This is a cool looking gimmick of weapon. It is OK as a weapon but not that good.

Added the OKC-3S Bayonet, a compromise between the army and national guard bayonets. It's Good as fighting knives go, but not as great as the old m7 bayonet.

Added the KA-BAR for a hunting knife that could do fighting knife things, without being a bayonet.

Turned the RM42 into (one of) the knives it was based on, the infamous Fairbairn Sykes knife.

Removed Bayonets from Garter sheaths??? Put more sensible weapons there where appropriate - usually the push dagger, sometimes the FS knife for assassination type knives.

Added/Adjusted spawns.

Removed all reach attack fire modes from bayonets.

While I did not intend to do a proper re-balance of knife damages/ DPS, I did notice some outliers - the sharp Kirpan 
 and gimmick machete overperformed, the baselard underperformed. The Hunting knife, now a skinner-type model, also overperformed and the RM42, now a real knife, overperformed. I made adjustments that I felt were conservative. I believe that knives overall need balancing but leave that to someone more capable.
 
 Here's a bar graph of the new DPS values shown in the item examine UI window.
image

#### Describe alternatives you've considered

Leaving all DPS's as they were with minimal adjustments.
Deleting the RM42.

#### Testing

Loaded games. Tested spawns (getting the resort one done was tricky.) Tested Bayonet reach. Checked that all the knives still fit in sheathes. Fewer things should probably fit in garter sheathes. 
#### Additional context

<https://imgur.com/a/y2ku0vo>Volume measurements here. I may have eyeballed widths here and there but overall I consider the volumes an improvement.

Table of patterns used
| Knife Name             | IRL Knife                                   |
| ---------------------- | ------------------------------------------- |
| Machete                | 1-18 Machete (OKC)                          |
| Baselard               | Arms-Armor Baselard Dagger, 14-15th Century |
| Hunting Knife          | Buck 103                                    |
| Survival Knife         | BUD K Rambo I Stallone Signature Knife      |
| Machete Multitool      | Camillus Carnivore Z 8"                     |
| Punch Dagger           | Cold Steel Drop Forged Push Knfe            |
| Karambit               | Cold Steel Steel Tiger                      |
| RM42                   | Fairbairn Sykes Commando Knife              |
| KA-BAR                 | KABAR knife                                 |
| Kirpan                 | Khalsa Raj G10 Kirpan -medium               |
| Sword Bayonet          | M1917 Sword Bayonet                         |
| Kukri          | Cold Steel Kukri/Khukri House Khukri 2022                         |
| Trench Knife           | M1918 Trench Knife                          |
| National Guard Bayonet | M7 Bayonet (OKC)                            |
| Army Bayonet           | M9 Bayonet (OKC)                            |
| USMC Bayonet           | OKC-3S  Bayonet (OKC)                       |
| Tanto                  | True Katana Shirasaya Tanto                 |

### CODE REVIEW

**Comment by TheShadowFerret** [comment]
If bayonets no longer give reach attack, why do we still limit pistols from having them 🤔

**Comment by fairyarmadillo** [comment]
You may want to give the kukri a look while you're at it. Its tool qualities are all about right but its DPS may now significantly outperform some of these blades. It's currently 450g but does 3 bash and 23 cut per swing.

I think the best thing to do would be to treat it like a slightly scaled down machete. Its forward angle and the fact that it's meant for combat might make it slightly better for its size, but it's definitely got less chopping power due to the reduced length and weight.

File: data/json/items/melee/swords_and_blades.json [hunk 1]
```diff
@@ -295,7 +326,51 @@
...
+      "install_time": "5 s"
+    },
+    "extend": { "flags": [ "PUMP_RAIL_COMPATIBLE" ] }
+  },
+  {
+    "id": "knife_combat",
+    "type": "TOOL",
+    "category": "weapons",
+    "name": { "str": "National Guard bayonet" },
+    "description": "A cold war era Army bayonet that has only partially been replaced by newer models.  It has an apocalypse-proof full-tang and a double-edged high carbon steel blade terminating in a spear point.  The no-frills design makes for both a good fighting knife and bayonet, a fact that was appreciated when reserve stocks of these were issued out to National Guardsmen.",
```
Comment (TheShadowFerret) [inline data/json/items/melee/swords_and_blades.json:359]:
```suggestion
    "description": "A Cold War era Army bayonet that has only partially been replaced by newer models.  It has an apocalypse-proof full-tang and a double-edged high carbon steel blade terminating in a spear point.  The no-frills design makes for both a good fighting knife and bayonet, a fact that was appreciated when reserve stocks of these were issued out to National Guardsmen.",
```

**Comment by tenmillimaster** [comment]
> If bayonets no longer give reach attack, why do we still limit pistols from having them(even w/ the modified version) 🤔

I think the answer to that question is just the limited space available for attaching the bayonet, without using a stubby specialized bayonet.

I think it'd be better to control this just via gun mod slots rather than by pistol vs non pistol.

**Comment by tenmillimaster** [comment]
> You may want to give the kukri a look while you're at it. Its tool qualities are all about right but its DPS may now significantly outperform some of these blades. It's currently 450g but does 3 bash and 23 cut per swing.
> 
> I think the best thing to do would be to treat it like a slightly scaled down machete. Its forward angle and the fact that it's meant for combat might make it slightly better for its size, but it's definitely got less chopping power due to the reduced length and weight.

I'll have a look and do an audit on its volume and weight as well.

File: data/json/items/melee/swords_and_blades.json [hunk 2]
```diff
@@ -295,7 +340,51 @@
...
+      "install_time": "5 s"
+    },
+    "extend": { "flags": [ "PUMP_RAIL_COMPATIBLE" ] }
+  },
+  {
+    "id": "knife_combat",
+    "type": "TOOL",
+    "category": "weapons",
+    "name": { "str": "National Guard bayonet" },
+    "description": "A Cold War era Army bayonet that has only partially been replaced by newer models.  It has an apocalypse-proof full-tang and a double-edged high carbon steel blade terminating in a spear point.  The no-frills design makes for both a good fighting knife and bayonet, a fact that was appreciated when reserve stocks of these were issued out to National Guardsmen.",
```
Comment (harakka) [inline data/json/items/melee/swords_and_blades.json:373]:
```suggestion
    "description": "A Cold War era Army bayonet that has only partially been replaced by newer models.  It has an apocalypse-proof full-tang, double-edged high carbon steel blade terminating in a spear point.  The no-frills design makes for both a good fighting knife and bayonet, a fact that was appreciated when reserve stocks of these were issued out to National Guardsmen.",
```

File: data/json/items/melee/swords_and_blades.json [hunk 3]
```diff
@@ -304,18 +393,18 @@
...
+    "melee_damage": { "stab": 22 }
   },
   {
     "id": "knife_combat_mod",
     "copy-from": "knife_combat",
     "type": "TOOL",
-    "name": { "str": "modified combat knife", "str_pl": "modified combat knives" },
-    "description": "A military combat knife.  It is light and extremely sharp, and could be deadly in either the right hands or when attached as a bayonet.  This one was modified and customized to mount on pretty much any weapon other than pistols, if you so want.",
+    "name": { "str": "modified National Guard bayonet" },
+    "description": "A cold war era Army bayonet that has only partially been replaced by newer models.  It has an apocalypse-proof full-tang and a double-edged blade of high carbon steel terminating in a spear point.  The no-frills design makes for both a good fighting knife and bayonet, a fact that was appreciated when reserve stocks of these were issued out to National Guardsmen.  This one was modified and customized to mount on pretty much any weapon other than pistols, if you so want.",
```
Comment (harakka) [inline data/json/items/melee/swords_and_blades.json:403]:
```suggestion
    "description": "A cold war era Army bayonet that has only partially been replaced by newer models.  It has an apocalypse-proof full-tang, double-edged blade of high carbon steel terminating in a spear point.  The no-frills design makes for both a good fighting knife and bayonet, a fact that was appreciated when reserve stocks of these were issued out to National Guardsmen.  This one was modified and customized to mount on pretty much any weapon other than pistols, if you so want.",
```

File: data/json/items/melee/swords_and_blades.json [hunk 4]
```diff
@@ -340,7 +429,28 @@
...
     "weapon_category": [ "KNIVES" ],
-    "melee_damage": { "cut": 16 }
+    "melee_damage": { "cut": 12 }
+  },
+  {
+    "id": "knife_KABAR",
+    "type": "TOOL",
+    "category": "weapons",
+    "name": { "str": "KA-BAR knife", "str_pl": "KA-BAR knives" },
+    "description": "The KA-BAR was developed by the United States Marine Corps in 1942 by modifying hunting knife designs to be more soldier-proof.  The handle is made of stacked leather washers, and the blade is an easy to sharpen high carbon steel, making for a popular fighting, hunting and utility knife.  The blade is marked 'USMC'.",
```
Comment (harakka) [inline data/json/items/melee/swords_and_blades.json:439]:
```suggestion
    "description": "The KA-BAR was developed by the United States Marine Corps in 1942 by modifying hunting knife designs to be more soldier-proof.  The handle is made of stacked leather washers, and the blade is made of easy to sharpen high carbon steel, making for a popular fighting, hunting and utility knife.  The blade is marked 'USMC'.",
```

**Review by harakka — COMMENTED** [review]
Some copyedits.

**Comment by Karol1223** [comment]
Could you alter the affected special attacks for ferals as well? There's at least one that was using a bayonet and one that was using a combat knife.

**Comment by Terrorforge** [comment]
>No bayonet makes any gun long enough to get reach attack

I have to be a pedant, because this is not strictly true. The flintlock musket is 1.46m, and with a sword bayonet (40cm) even some of the longer modern guns are at least _almost_ there.

I don't think there's a great way to implement that, though. Better to err on the side of caution. Maybe make one of those side-mounted triangular bayonets specifically for the flintlock, or just give it reach naturally.

**Comment by tenmillimaster** [comment]
> > No bayonet makes any gun long enough to get reach attack
> 
> I have to be a pedant, because this is not strictly true. The flintlock musket is 1.46m, and with a sword bayonet (40cm) even some of the longer modern guns are at least _almost_ there.
> 
> I don't think there's a great way to implement that, though. Better to err on the side of caution. Maybe make one of those side-mounted triangular bayonets specifically for the flintlock, or just give it reach naturally.

I intend to do this, actually, add a bayonet targeting specfic guns that do give you reach.

**Comment by tenmillimaster** [comment]
> You may want to give the kukri a look while you're at it. Its tool qualities are all about right but its DPS may now significantly outperform some of these blades. It's currently 450g but does 3 bash and 23 cut per swing.
> 
> I think the best thing to do would be to treat it like a slightly scaled down machete. Its forward angle and the fact that it's meant for combat might make it slightly better for its size, but it's definitely got less chopping power due to the reduced length and weight.

image

Done.

**Comment by tenmillimaster** [comment]
> Could you alter the affected special attacks for ferals as well? There's at least one that was using a bayonet and one that was using a combat knife.

Alright done! TY for the suggestion.

**Comment by Terrorforge** [comment]
> > > No bayonet makes any gun long enough to get reach attack
> > 
> > 
> > I have to be a pedant, because this is not strictly true. The flintlock musket is 1.46m, and with a sword bayonet (40cm) even some of the longer modern guns are at least _almost_ there.
> > I don't think there's a great way to implement that, though. Better to err on the side of caution. Maybe make one of those side-mounted triangular bayonets specifically for the flintlock, or just give it reach naturally.
> 
> I intend to do this, actually, add a bayonet targeting specfic guns that do give you reach.

Worth mentioning that this would still give them reach even after you cut down the stock/barrel. Not sure if there's a great way around that, or if it's one of those weird thigns about the weapon mod system you just have to accept. And that's how it currently works, anyway.

**Comment by tenmillimaster** [comment]
> > > > No bayonet makes any gun long enough to get reach attack
> > > 
> > > 
> > > I have to be a pedant, because this is not strictly true. The flintlock musket is 1.46m, and with a sword bayonet (40cm) even some of the longer modern guns are at least _almost_ there.
> > > I don't think there's a great way to implement that, though. Better to err on the side of caution. Maybe make one of those side-mounted triangular bayonets specifically for the flintlock, or just give it reach naturally.
> > 
> > 
> > I intend to do this, actually, add a bayonet targeting specfic guns that do give you reach.
> 
> Worth mentioning that this would still give them reach even after you cut down the stock/barrel. Not sure if there's a great way around that, or if it's one of those weird thigns about the weapon mod system you just have to accept. And that's how it currently works, anyway.

I have something in mind. Out of scope for this PR, though, I let that grow enough.

**Comment by ehughsbaird** [comment]
It looks like the DPS tests need to be made aware of the changes.
https://github.com/CleverRaven/Cataclysm-DDA/blob/157ab107c9c9c964a7ae6109b854021ca4ed0423/data/mods/TEST_DATA/expected_dps_data/knives_dps.json#L7-L32

File: data/mods/TEST_DATA/expected_dps_data/knives_dps.json [hunk 5]
```diff
@@ -3,30 +3,29 @@
...
+      "kirpan_cheap": 2.809,
       "kris": 13.5,
       "kris_fake": 1.811,
       "knife_meat_cleaver": 13.97,
       "pockknife": 4.5,
       "primitive_knife": 2.5,
-      "mc_knife_rambo": 10.47,
-      "knife_rm42": 30.11,
+      "mc_knife_rambo": 9.415,
+      "knife_rm42": 17.580,
```
[the remark this answers is not in the document]
Comment (tenmillimaster) [reply data/mods/TEST_DATA/expected_dps_data/knives_dps.json:23]: Gosh GH, I'll have you know that trailing zeroes are very important to scientists such as ME. /J

**Comment by tenmillimaster** [comment]
> It looks like the DPS tests need to be made aware of the changes.
> 
> https://github.com/CleverRaven/Cataclysm-DDA/blob/157ab107c9c9c964a7ae6109b854021ca4ed0423/data/mods/TEST_DATA/expected_dps_data/knives_dps.json#L7-L32

Updated! Thanks. Now it's just failing on ... wield times?

**Comment by ehughsbaird** [comment]
The correct value is 105. If you want within one move like the comment has, I'd change the condition to `knife_obtain_cost == Approx( 105 ).margin( 1 )`.

**Comment by Mister-Cultist** [comment]
sword bayonets should still give reach, atleast on longer guns.

**Pull request merged**